### PR TITLE
Subtract Templates for MOSS

### PR DESCRIPTION
**Pull request by Sumukh**
Since MOSS doesn't seem to be handling our templates well - we should try removing templates manually.

### CODE REVIEW

File: server/jobs/moss.py [hunk 1]
```diff
@@ -72,6 +88,13 @@ def submit_to_moss(moss_id=None, file_regex=".*", assignment_id=None, language=N
             for file in contents:
                 if file == 'submit':  # ignore fake file from ok-client
                     continue
+                if subtract_template and file in assign.files:
+                    # Compare to template and only include lines that new
+                    template, source = assign.files[file], contents[file]
+                    d = difflib.Differ()
```
Comment (Sumukh) [inline server/jobs/moss.py]: I think we should make difflib ignore spacing and new lines (Extra indentation, Tabs v Spaces ...)  - we can do that with the two args to `Differ` (Junk Char being one of them)

File: server/forms.py [hunk 2]
```diff
@@ -446,6 +445,8 @@ class MossSubmissionForm(BaseForm):
     file_regex = StringField('Regex for submitted files', default='.*',
                              validators=[validators.required()])
     language = SelectField('Language', choices=[(pl, pl) for pl in COMMON_LANGUAGES])
+    subtract_template = BooleanField('Subtract Template', default=False,
```
Comment (knrafto) [inline server/forms.py]: Should the default be True?

File: server/jobs/moss.py [hunk 3]
```diff
@@ -57,6 +59,20 @@ def submit_to_moss(moss_id=None, file_regex=".*", assignment_id=None, language=N
         match_pattern = re.compile(file_regex)
         ignored_files = set()
 
+        template_files = []
+        for template in assign.files:
+            dest = "{}/{}".format(tmp_dir, template)
```
Comment (knrafto) [inline server/jobs/moss.py]: Use `os.path.join`

File: server/jobs/moss.py [hunk 4]
```diff
@@ -57,6 +59,20 @@ def submit_to_moss(moss_id=None, file_regex=".*", assignment_id=None, language=N
...
+            with open(dest, 'w') as f:
+                f.write(assign.files[template])
+            template_files.append(template)
```
Comment (knrafto) [inline server/jobs/moss.py]: template_files is just list(assign.files). You could pass assign.files to the ''.join below directly

File: server/jobs/moss.py [hunk 5]
```diff
@@ -72,6 +88,13 @@ def submit_to_moss(moss_id=None, file_regex=".*", assignment_id=None, language=N
             for file in contents:
                 if file == 'submit':  # ignore fake file from ok-client
                     continue
+                if subtract_template and file in assign.files:
+                    # Compare to template and only include lines that new
+                    template, source = assign.files[file], contents[file]
+                    d = difflib.Differ()
+                    diff = d.compare(template.splitlines(True), source.splitlines(True))
```
Comment (knrafto) [inline server/jobs/moss.py]: splitlines(keepends=True) instead of splitlines(True) is clearer

**Comment by knrafto** [comment]
I'm assuming you've tested it already, so besides a few style nits I think it's good to go

File: server/forms.py (hunk 2 above)
Comment (Sumukh) [inline server/forms.py]: I think we should leave it off by default (from my testing) - It just seems really good at catching last year's template. I have to evaluate it some more.

File: server/jobs/moss.py [hunk 6]
```diff
@@ -57,21 +59,44 @@ def submit_to_moss(moss_id=None, file_regex=".*", assignment_id=None, language=N
...
+
         for backup in backup_query:
             # Write file into file
             file_contents = [m for m in backup.messages if m.kind == 'file_contents']
             if not file_contents:
                 logger.info("{} didn't have any file contents".format(backup.hashid))
                 continue
             contents = file_contents[0].contents
-            dest_dir = "{}/{}/".format(tmp_dir, backup.hashid)
+            dest_dir = "{}/".format(os.path.join(tmp_dir, backup.hashid))
```
Comment (knrafto) [inline server/jobs/moss.py]: I think this `format` is unnecessary? Maybe you need `os.path.isdir`

**Pull request merged**